File: bindings/python/examples/bench.py
#!/usr/bin/env python3
"""Python throughput benchmark, mirrors examples/bench.cpp."""
from __future__ import annotations

import sys
import time

from nyxstone_tricore_gcc import NyxstoneTricoreGCC


INSNS = [
    "mov %d4, %d5", "add %d4, %d5", "sub %d4, %d5", "nop", "ret",
    "mov.aa %a4, %a5", "ld.w %d4, [%a4]", "st.w [%a4], %d4",
    "and %d4, %d5", "or %d4, %d5",
]


def make_pkg(n: int) -> str:
    return "; ".join(INSNS[i % len(INSNS)] for i in range(n))


def run_for(secs: float, fn) -> float:
    for _ in range(3):
        fn()
    t0 = time.perf_counter()
    count = 0
    while True:
        for _ in range(10):
            fn()
            count += 1
        e = time.perf_counter() - t0
        if e >= secs:
            return count / e


def fmt(v: float) -> str:
    if v >= 1e6: return f"{v/1e6:>8.2f} M"
    if v >= 1e3: return f"{v/1e3:>8.2f} k"
    return f"{v:>10.2f}"


def main() -> None:
    secs = float(sys.argv[1]) if len(sys.argv) > 1 else 1.0
    nx = NyxstoneTricoreGCC()
    print(f"NyxstoneTricoreGCC (Python) bench, target {secs}s/measurement")
    for pkg in (1, 10):
        text = make_pkg(pkg)
        b = nx.assemble(text)
        r_asm = run_for(secs, lambda: nx.assemble(text))
        r_dis = run_for(secs, lambda: nx.disassemble_to_instructions(b))
        print(f"  Package {pkg} ({pkg} insns, {len(b)} bytes)")
        print(f"    assemble    : {fmt(r_asm)} ops/s   {fmt(r_asm*pkg)} insns/s")
        print(f"    disassemble : {fmt(r_dis)} ops/s   {fmt(r_dis*pkg)} insns/s")


if __name__ == "__main__":
    main()

File: bindings/rust/examples/bench.rs
// Throughput benchmark for the Rust binding, mirrors examples/bench.cpp.
// Usage: cargo run --release --example bench -- [seconds]
use nyxstone_tricore_gcc::NyxstoneTricoreGCC;
use std::time::Instant;

const INSNS: [&str; 10] = [
    "mov %d4, %d5", "add %d4, %d5", "sub %d4, %d5", "nop", "ret",
    "mov.aa %a4, %a5", "ld.w %d4, [%a4]", "st.w [%a4], %d4",
    "and %d4, %d5", "or %d4, %d5",
];

fn make_pkg(n: usize) -> String {
    let mut s = String::new();
    for i in 0..n {
        if i > 0 { s.push_str("; "); }
        s.push_str(INSNS[i % INSNS.len()]);
    }
    s
}

fn run_for<F: FnMut()>(secs: f64, mut f: F) -> f64 {
    for _ in 0..3 { f(); }
    let t0 = Instant::now();
    let mut count = 0u64;
    loop {
        for _ in 0..10 { f(); count += 1; }
        let e = t0.elapsed().as_secs_f64();
        if e >= secs { return (count as f64) / e; }
    }
}

fn fmt(v: f64) -> String {
    if v >= 1e6 { format!("{:>8.2} M", v / 1e6) }
    else if v >= 1e3 { format!("{:>8.2} k", v / 1e3) }
    else { format!("{v:>10.2}") }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let secs = std::env::args().nth(1)
        .and_then(|s| s.parse::<f64>().ok())
        .unwrap_or(1.0);

    let nx = NyxstoneTricoreGCC::new()?;
    println!("NyxstoneTricoreGCC (Rust) bench, target {secs}s/measurement");

    for &pkg in &[1usize, 10] {
        let text = make_pkg(pkg);
        let bytes = nx.assemble(&text, 0, &[])?;
        let r_asm = run_for(secs, || { let _ = nx.assemble(&text, 0, &[]); });
        let r_dis = run_for(secs, || { let _ = nx.disassemble_to_instructions(&bytes, 0, 0); });
        println!("  Package {pkg} ({pkg} insns, {} bytes)", bytes.len());
        println!("    assemble    : {} ops/s   {} insns/s",
                 fmt(r_asm), fmt(r_asm * pkg as f64));
        println!("    disassemble : {} ops/s   {} insns/s",
                 fmt(r_dis), fmt(r_dis * pkg as f64));
    }
    Ok(())
}
